### PR TITLE
Support FullScreen & Media

### PR DESCRIPTION
**Pull request by mshin-wolvic**
- Add TabMediaSessionObserver for the media control
- Add TabWebContentsDelegate for fullscreen

Precursor CL is Igalia/wolvic-chromium#7

### CODE REVIEW

**Review by duswnchl — COMMENTED** [review]
If it is possible, what about to separate this PR in two parts?  For example, `Support fullscreen for Chromium backend` and `Support media for Chromium backend`. It could be 2 PRs or 2 clearly separated commits in the current PR.

**Comment by mshin-wolvic** [comment]
> If it is possible, what about to separate this PR in two parts? For example, `Support fullscreen for Chromium backend` and `Support media for Chromium backend`. It could be 2 PRs or 2 clearly separated commits in the current PR.

Good question. I also thought about dividing this into two CLs, but after it's on full screen, it's able to connect the media controller so that it can function (because we don't have a unit test code at the moment) and the full screen will work only when the media observer is implemented. So I created one because of the high interdependence.

**Review by jdapena — APPROVED** [review]

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java
Comment (svillar) [inline app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java:1]: The changes to this file deserve its own commit. It's a bugfix by itself

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java [hunk 1]
```diff
@@ -54,7 +59,7 @@ public SessionImpl(@Nullable WSessionSettings settings) {
 
     private void init() {
         mTextInput = new TextInputImpl(this);
-        mPanZoomCrontroller = new PanZoomCrontrollerImpl(this);
+        mPanZoomController = new PanZoomControllerImpl(this);
```
Comment (svillar) [inline app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java:62]: Same here, the panZoomController typo can be a separate commit, it's unrelated to the rest of the changes

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabMediaSessionObserver.java [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+import org.chromium.services.media_session.MediaImage;
+import org.chromium.services.media_session.MediaMetadata;
+import org.chromium.services.media_session.MediaPosition;
+
+import java.util.List;
+
+import org.chromium.media_session.mojom.MediaSessionAction;
+
+public class TabMediaSessionObserver extends MediaSessionObserver implements MediaImageCallback {
+    private static final int UPDATE_POSITION_TIME_MS = 500;
```
Comment (svillar) [inline app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabMediaSessionObserver.java:33]: Where does this 500ms come from?

**Review by svillar — CHANGES_REQUESTED** [review]
Looking great, I would just like to split the patch in 3 commits (see comments)

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabMediaSessionObserver.java (hunk 2 above)
Comment (mshin-wolvic) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabMediaSessionObserver.java:33]: This comes from my prediction. :)
Chrome browser uses a 250 ms period and manages a progress bar based on the milliseconds unit.
Wolvic manages it based on the seconds unit, so I thought it is enough to update the position with a 500 ms period. But I don't have a strong opinion here.

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java (hunk 1 above)
Comment (mshin-wolvic) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java]: Removed and fixed this typo issue by Zakhar.

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java
Comment (mshin-wolvic) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/DisplayImpl.java:1]: branched out #607

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabMediaSessionObserver.java [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+    private WImage mMediaImage;
+    private MediaPosition mMediaPosition;
+    private boolean mIsActive = false;
+    private boolean mIsSuspended = false;
+    private boolean mRunUpdatingPositionTask = false;
+
+    public TabMediaSessionObserver(@NonNull WebContents webContents, @NonNull SessionImpl session) {
+        super(MediaSession.fromWebContents(webContents));
+
+        mWebContents = webContents;
```
Comment (duswnchl) [inline app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabMediaSessionObserver.java:49]: Should we keep the reference `mWebContents`? Because this private reference doesn't seem to use in this class except here.

Comment (mshin-wolvic) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabMediaSessionObserver.java]: Done.

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java [hunk 4]
```diff
@@ -43,6 +45,9 @@ public class SessionImpl implements WSession {
     PanZoomControllerImpl mPanZoomController;
     private ActivityWindowAndroid mWindowAndroid;
     private IntentRequestTracker mIntentRequestTracker;
+    private Tab mTab = null;
```
Comment (svillar) [inline app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java:48]: No need for null, that's the default for Java objects

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabMediaSessionObserver.java [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+    }
+
+    private void updatePosition() {
+        if (mMediaPosition == null || !mIsActive)
+            return;
+
+        assert mMediaSession != null;
+
+        long now = SystemClock.elapsedRealtime();
+        long rebased_position = mMediaPosition.getPosition()
```
Comment (svillar) [inline app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabMediaSessionObserver.java:213]: nit: rebase_position -> rebasedPosition

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsDelegate.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+       if (delegate != null && !reverse) {
+           delegate.onFocusRequest(mSession);
+       }
+       return false;
+   }
+
+    @Override
+    public void enterFullscreenModeForTab(boolean prefersNavigationBar, boolean prefersStatusBar) {
+        @Nullable WSession.ContentDelegate delegate = mSession.getContentDelegate();
+        if (delegate != null) {
```
Comment (svillar) [inline app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsDelegate.java:34]: In general we prefer early returns to avoid extra indentation as you perfectly did in most of the cases.. So in this case
```
if (delegate==null)
 return;
```

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsDelegate.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+            mIsFullscreen = true;
+            delegate.onFullScreen(mSession, true);
+        }
+    }
+
+    @Override
+    public void exitFullscreenModeForTab() {
+        @Nullable WSession.ContentDelegate delegate = mSession.getContentDelegate();
+        if (delegate != null) {
```
Comment (svillar) [inline app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsDelegate.java:43]: Ditto early return

**Review by svillar — COMMENTED** [review]
Awesome!. I just have a few very minor nits and then we could land

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabMediaSessionObserver.java (hunk 5 above)
Comment (mshin-wolvic) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabMediaSessionObserver.java]: Done

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java (hunk 4 above)
Comment (mshin-wolvic) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/SessionImpl.java]: Done

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabMediaSessionObserver.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mshin-wolvic) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabMediaSessionObserver.java:34]: Added.

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsDelegate.java (hunk 6 above)
Comment (mshin-wolvic) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsDelegate.java]: Done

File: app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsDelegate.java (hunk 7 above)
Comment (mshin-wolvic) [reply app/src/common/chromium/com/igalia/wolvic/browser/api/impl/TabWebContentsDelegate.java]: Done

**Review by svillar — APPROVED** [review]
Awesome, could you please combine the review commits with the 3 previous patches so that we only have 3 patches to merge?
